sys-boot/plymouth/plymouth-24.004.60.ebuild

```diff
@@ -0,0 +1,118 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+inherit meson readme.gentoo-r1 flag-o-matic
+
+SRC_URI="https://dev.gentoo.org/~aidecoe/distfiles/${CATEGORY}/${PN}/gentoo-logo.png"
+
+if [[ ${PV} == 9999 ]]; then
+	inherit git-r3
+	EGIT_REPO_URI="https://gitlab.freedesktop.org/plymouth/plymouth"
+else
+	SRC_URI="${SRC_URI} https://www.freedesktop.org/software/plymouth/releases/${P}.tar.xz"
+	KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~ia64 ~loong ~ppc ~ppc64 ~riscv ~sparc ~x86"
+fi
+
+DESCRIPTION="Graphical boot animation (splash) and logger"
+HOMEPAGE="https://cgit.freedesktop.org/plymouth/"
+
+LICENSE="GPL-2+"
+SLOT="0"
+IUSE="debug +drm +gtk +pango selinux freetype +split-usr +udev doc upstart-monitoring systemd"
+
+BDEPEND="
+	app-text/docbook-xsl-stylesheets
+	>=dev-build/meson-0.62
+	dev-libs/libxslt
+	virtual/pkgconfig
+"
+
+DEPEND="
+	dev-libs/libevdev
+	>=media-libs/libpng-1.2.16:=
+	virtual/libc
+	x11-libs/libxkbcommon
+	x11-misc/xkeyboard-config
+	drm? ( x11-libs/libdrm )
+	elibc_musl? ( sys-libs/rpmatch-standalone )
+	freetype? ( media-libs/freetype:2 )
+	gtk? (
+		dev-libs/glib:2
+		x11-libs/cairo
+		>=x11-libs/gtk+-3.14:3[X]
+	)
+	pango? (
+		x11-libs/cairo
+		>=x11-libs/pango-1.21
+	)
+	systemd? ( sys-apps/systemd )
+	udev? ( virtual/libudev )
+	upstart-monitoring? (
+		sys-apps/dbus
+		sys-libs/ncurses
+	)
+"
+
+RDEPEND="${DEPEND}
+	selinux? ( sec-policy/selinux-plymouthd )
+	udev? ( virtual/udev )
+"
+
+DOC_CONTENTS="
+	Follow the following instructions to set up Plymouth:\n
+	https://wiki.gentoo.org/wiki/Plymouth#Configuration
+"
+
+src_prepare() {
+	use elibc_musl && append-ldflags -lrpmatch
+	default
+}
+
+src_configure() {
+	local emesonargs=(
+		--localstatedir "${EPREFIX}"/var
```

Comment:
This I think should be:
```
-Drunstatedir=/var
```

Currently the build produces QA warnings, and this does not look like the correct style for meson options.

```
* QA Notice: The ebuild is installing to one or more unexpected directories:
*
*   /run
*
* Please fix the ebuild to use correct FHS/Gentoo policy paths.
* QA Notice: This ebuild installs into paths that should be created at runtime.
*  To fix, simply do not install into these directories.  Instead, your package
*  should create dirs on the fly at runtime as needed via init scripts/etc...
*
*   run
*   run/plymouth
* QA Notice: One or more empty directories installed to /var:
*
*   /var/lib
*   /var/lib/plymouth
*   /var/spool
*   /var/spool/plymouth
```

Reply:
I added `--localstatedir "${EPREFIX}"/var` since it would go to `/var/lib/plymouth/plymouth` without it, and also `localstatedir` was supplied in the previous ebuild 22.02.122-r1 (not on meson)

Whilst adding `-Drunstatedir=/var` removes the notice for /run, the /var notices are there, I also believe it is not appropriate to set `-Drunstatedir=/var` since in the plymouth dracut pretrigger it creates and looks for /run for plymouth (though I haven't tested or looked into what would happen with it set to /var).

I believe there should be another solution, though im not really sure what it'd be. We could try delete the /run/plymouth directory after since it is empty when created, and also it should created by the initramfs before running - so maybe we can remove it somehow? Though i'm not sure how to remove it properly in the ebuild.

For the /var notices they were still there on the previous ebuilds, should I just `keepdir` them? However, I'm unsure on what they are being used for

Reply:
> I also believe it is not appropriate to set `-Drunstatedir=/var` since in the plymouth dracut pretrigger it creates and looks for /run for plymouth (though I haven't tested or looked into what would happen with it set to /var).

That is in the initramfs, there is no need to have this on the rootfs right? The current ebuild does not install anything outside of FHS and is functional. The current ebuild also has a comment about deleting `/var/run` which may be related.

> For the /var notices they were still there on the previous ebuilds, should I just `keepdir` them? However, I'm unsure on what they are being used for

`keepdir` them only if they are actually used.

Reply:
>That is in the initramfs, there is no need to have this on the rootfs right? The current ebuild does not install anything outside of FHS and is functional. The current ebuild also has a comment about deleting /var/run which may be related.

Yeah your right about the initramfs my bad :sweat_smile:, the old build system probably sets `runstatedir` as `/var/run`. Though it probably doesn't need to be changed now anyways since it can stay as `/run` but be deleted at the end of the ebuild - see below on my reasoning (so it's probably better to leave it as `/run` since `/var/run` is just symlinked to `/run` due to FHS 3.0 anyways)


Right, so I've just had a quick look at the source, and from what I can tell:

`/var/lib/plymouth` is the time directory (which stores boot-duration and shutdown-duration), it is created empty during compilation, however we can remove this directory [since it is created at runtime if it doesn't already exist](https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/24.004.60/src/main.c?ref_type=tags#L1432)

`/var/spool/plymouth` is used to place error logs when there's errors, I think this directory can be left using `keepdir` since there's no check to ensure it is created before writing the log (not 100% sure)? (the file is linked to `/var/spool/plymouth/boot.log` using https://linux.die.net/man/3/link - which I'm not sure if it fails if the destination directory doesn't exist): error logged in these two functions https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/24.004.60/src/main.c?ref_type=tags#L779, which [ends up using `link` here](https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/24.004.60/src/libply/ply-utils.c#L651)

And finally, `/run/plymouth` is also [created at runtime similar to `/var/lib/plymouth` so it can also be deleted](https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/24.004.60/src/main.c?ref_type=tags#L2125)

It's probably best to double check these though.

Also if you think that its best to set `-Drunstatedir=/var` still then just let me know :sweat_smile:!

Reply:
> Also if you think that its best to set `-Drunstatedir=/var` still then just let me know 😅!

I suppose it doesn't really matter then since it is symlinked. For simplicity lets then not set this and use the default as it is set by the build system.

> It's probably best to double check these though.

Thanks for digging into this. Since the directories are used, I think `keepdir` is the best solution. This avoids surprises if something (in plymouth or some other package) in the future changes and fails when this directory has (not yet) been created.

Reply:
> This I think should be:
> 
> ```
> -Drunstatedir=/var
> ```
> Currently the build produces QA warnings, and this does not look like the correct style for meson options.

In meson.eclass, the default meson arguments includes:

```bash
        MESONARGS+=(
                --libdir "$(get_libdir)"
                --localstatedir "${EPREFIX}/var/lib"
                --prefix "${EPREFIX}/usr"
                --sysconfdir "${EPREFIX}/etc"
```